_posts/2024-12-16-edge-computing-tech.md

```diff
@@ -0,0 +1,59 @@
+It is an extension of sorts of IoT Hub. Iot Edge containing two docker containers that you set up locally. 
+The first container, **edgeAgent**, monitors the local workloads and automatically downloads and launches the correct version if the deployment manifest (a JSON file in Azure) is changed. The second container, **edgeHub**, handles all communication, acting as a message broker between custom IoT edge modules, the device, the cloud, and even other edge devices (this is called the gateway pattern).
+A useful feature of the edgeHub is that it allows for message retention —it can store messages locally when offline, and when the device reconnects, it sends all missed D2C messages to IoT Hub, which is a very nice out-of-the-box feature.
+With this system you can run almost anything on the edge, within the limits of your hardware that is. There are ready made marketplace modules like Stream Analytics and OPC UA clients as well as modules for Azure AI services. You can also deploy custom workloads by adding some IoT edge-specific code to your app and containerizing it. 
+Below is a reference architecture for running a data collection and analysis solution on the edge, for more info check this guide: [Data Storage Edge.](https://learn.microsoft.com/en-us/azure/architecture/solution-ideas/articles/data-storage-edge)
+![Data storage and analysis at the edge](/img/2024-edge-computing/iot_edge.png)
+*Data storage and analysis at the edge, https://learn.microsoft.com/en-us/azure/architecture/solution-ideas/articles/data-storage-edge
+ 
+Again, Aws has a quite similar product called **Aws IoT Greengrass**. 
+
+While IoT Edge and Greengrass are both excellent products, a word of caution. At least for IoT Edge these benefits come with the cost of a rather mediocre dev experience. Do you remember the “IoT edge-specific code” that I mentioned earlier? Yeah, that makes the application dependent on the edge runtime. Also, your CD pipelines need to do the deployments via the IoT hub and that opens another can of worms. 
+You can get around these things, and Microsoft even provides an edge runtime simulator for vsCode, so you do have the possibility to run the system locally, but the system is still very much tied to IoT Hub and the edge runtime. 
+I don’t have personal experience with Greengrass, perhaps it's all rainbows and sunshine with that, or maybe not. 
+
+# What about a very big things? 
+The solutions introduced until now are dealing with devices. They can be big machines - IoT Edge was used for modernizing a forest machine at Ponsse, but still, it is a quite small system. So, what to do if you want to connect an entire factory? For this purpose, we have **Azure IoT Operations**, or AIO for short. 
+This is a brand-new product that reached GA in the last week of November 2024. 
```

Comment:
What is GA? General Availability? Would spell it out and introduce the GA acronym for the remainder of the text

Reply:
Yes I should introduce all acronyms, will fix this